### ISSUE TITLE
Saving and Loading fine-tuned model

### ISSUE THREAD

**Issue by binbinxue** [huggingface/peft#1384]
A quick and stupid question. 
After going through a lot of tutorials (official and private), i find myself confused about saving a BitAndBytes and QLora fine-tuned model. 

In the tutorials, base models were loaded in 4bit using BitAndBytes quantization and QLora adapater is added for fine-tuning. 
I know that after the model.save_pretrained() only the adapter is saved.

But then afterwards, for inference people just load the full precision base model without the BitAndBytes quantization and add the adapter. Isn't this wrong? If the adapter was trained with Base Model loaded using BitAndBytes configuration then it's optimised for this quantized base model? Shouldn't the base model be loaded with same BitAndBytes configuration before adding the adapter for inference?

**Comment by binbinxue** [comment]
> Hi binbinxue,
> 
> ![image](https://private-user-images.githubusercontent.com/19851002/298846066-c9dc971c-4414-4b1b-86dd-44d06260942e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYwMDA1MjgsIm5iZiI6MTcwNjAwMDIyOCwicGF0aCI6Ii8xOTg1MTAwMi8yOTg4NDYwNjYtYzlkYzk3MWMtNDQxNC00YjFiLTg2ZGQtNDRkMDYyNjA5NDJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIzVDA4NTcwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5YjcwOTRjMzIzNmZiOWQ3MTdjMTE0ZDhhY2NlZGUxOTk3NjE0OWQ3ZmQxZTBhZWVlYjQ0YjFkZWQ3MTM1YTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7hkRiQrmdfFwyCjbjGxucq2SW2n9stCtyJJKR-0O5ns)
> 
> Excerpts from the [QLoRA](https://arxiv.org/abs/2305.14314) paper:
> 
> > QLoRA has one low-precision storage data type, in our case usually 4-bit, and one computation data type that is usually BFloat16. In practice, this means whenever a QLoRA weight tensor is used, we dequantize the tensor to BFloat16, and then perform a matrix multiplication in 16-bit.
> 
> > To summarize, QLoRA has one storage data type (usually 4-bit NormalFloat) and a computation data type (16-bit BrainFloat). We dequantize the storage data type to the computation data type to perform the forward and backward pass, but we only compute weight gradients for the LoRA parameters which use 16-bit BrainFloat.
> 
> The base model is dequantized to the computation data type which is in 16-bits during fine-tuning. So both the base model and adapters are in 16-bits when this happens.
> 
> That means that when you want to use the adapters for inference, quantizing the base model is not necessary.
> 
> Hope that clear things up.

Thanks for the reply, it's still confusing to me, if the base model was quantized during training (say load_in_4bit is true etc), won't there be a loss of precision after dequantization for computation (for the feedforward)? I would have thought the difference in bits will result in the difference in computed loss. Won't the adapter be optimized for the 4bit quantized model during training? 

after training if base model is using the full precision, won't the extra bit contribute to some small difference and these small differences accumulate to be big difference for the outputs?

**Comment by nikich340** [comment]
QLoRA ≠ training in 4-bit math
QLoRA = training LoRA on dequantized F16 weights, which contains some error compared to original F16 weights.
So while adapter is trained on "a bit mismatched weights" it can be definetely applied to original F16 weights, what will produce better results (my guess: diff in original vs dequantized model >> diff in adapter weights trained on original vs dequantized model).
